pkg/reconciler/source/reconciler.go

```diff
@@ -0,0 +1,136 @@
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package source
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/google/kf/pkg/apis/kf/v1alpha1"
+	kflisters "github.com/google/kf/pkg/client/listers/kf/v1alpha1"
+	"github.com/google/kf/pkg/reconciler"
+	"github.com/google/kf/pkg/reconciler/source/resources"
+	cbuild "github.com/knative/build/pkg/client/clientset/versioned/typed/build/v1alpha1"
+	buildlisters "github.com/knative/build/pkg/client/listers/build/v1alpha1"
+	"go.uber.org/zap"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+	"knative.dev/pkg/controller"
+	"knative.dev/pkg/logging"
+)
+
+// Reconciler reconciles an source object with the K8s cluster.
+type Reconciler struct {
+	*reconciler.Base
+
+	buildClient cbuild.BuildV1alpha1Interface
+
+	// listers index properties about resources
+	SourceLister kflisters.SourceLister
+	buildLister  buildlisters.BuildLister
+}
+
+// Check that our Reconciler implements controller.Reconciler
+var _ controller.Reconciler = (*Reconciler)(nil)
+
+// Reconcile is called by Kubernetes.
+func (r *Reconciler) Reconcile(ctx context.Context, key string) error {
+	logger := logging.FromContext(ctx)
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	original, err := r.SourceLister.Sources(namespace).Get(name)
+	switch {
+	case errors.IsNotFound(err):
+		logger.Errorf("source %q no longer exists\n", name)
+		return nil
+
+	case err != nil:
+		return err
+
+	case original.GetDeletionTimestamp() != nil:
+		return nil
+	}
+
+	// Don't modify the informers copy
+	toReconcile := original.DeepCopy()
+
+	// Reconcile this copy of the service and then write back any status
+	// updates regardless of whether the reconciliation errored out.
+	reconcileErr := r.ApplyChanges(ctx, toReconcile)
+	if equality.Semantic.DeepEqual(original.Status, toReconcile.Status) {
+		// If we didn't change anything then don't call updateStatus.
+		// This is important because the copy we loaded from the informer's
+		// cache may be stale and we don't want to overwrite a prior update
+		// to status with this stale state.
+
+	} else if _, uErr := r.updateStatus(namespace, toReconcile); uErr != nil {
+		logger.Warnw("Failed to update Source status", zap.Error(uErr))
+		return uErr
+	}
+
+	return reconcileErr
+}
+
+// ApplyChanges updates the linked resources in the cluster with the current
+// status of the source.
+func (r *Reconciler) ApplyChanges(ctx context.Context, source *v1alpha1.Source) error {
+	// Sync build
+	{
```

Comment:
What if this is a docker image rather than a buildpack build?

I think you could handle it in one of two ways:

1. short-circuit this
2. create a no-op build for docker images

Reply:
I did 2. since we might want to add validation steps/other modifications to the pipeline in the future.